### DISCUSSION TITLE
Switch existing service implementation to a custom one before usage

### DISCUSSION THREAD

**Discussion by lucian-torje-siemens** [eclipse-edc/Connector#1686] [Ideas]
Hello,
I would like to be able to register my own service before it gets used by other registered service  - e.g. CatalogService
It is possible to register my own service - the issue is that the other registered services using it, will still use the old one instead of my own. 

The main idea is to register a service with priority - meaning that those special services (components) get registered before the default ones and custom ones cannot be overwritten once created (and will be injected instead of the default ones in other components).

![image](https://user-images.githubusercontent.com/102717569/178917995-03e74b59-d48f-4d9b-89c9-1865333a9e63.png)

**Comment by ndr-brt** [comment]
Hi, the way to achieve that is to register the "default" one with the `@Provider(isDefault = true)` annotation, then the custom service can be registered by an extension.
Right now a lot of services would need to be set as default, feel free to open issue/pr about specific services.
Ref: https://eclipse-dataspaceconnector.github.io/DataSpaceConnector/#/developer/dependency_resolution?id=provide-quotdefaultsquot

**Reply by lucian-torje-siemens** [reply]
thank you @ndr-brt

**Reply by lucian-torje-siemens** [reply]
Hi @ndr-brt , when you have time, please review the code https://github.com/mindsphere/DataSpaceConnector-Fork/pull/37/files#diff-39fe69e76d10b7f7499a610c0ff72f3a5b56dc1c11f06cc5213477d8b786b1d2 - at the end I could not make the Provider annotation work so I returned to using the manual approach of registering plus overriding the register function due to the limitation or registering multiple times the same service.

**Reply by ndr-brt** [reply]
@lucian-torje-siemens , are the `ConnectorServiceImpl` and the `CatalogServiceImpl` that need to be provided by default on `IdsCoreServiceExtension`, then you could overwrite them with yours using the `@Provider` annotation (or registering them with the `registerService` plus the `@Provides` annotation), could you open an issue/PR about this?

**Reply by jimmarino** [reply]
> Hello, I would like to be able to register my own service before it gets used by other registered service - e.g. CatalogService It is possible to register my own service - the issue is that the other registered services using it, will still use the old one instead of my own.

As @ndr-brt mentioned, default services enable this. However, there is a broader question of why you need to replace the ConnectorService and CatalogService implementations. Can you please describe what you are trying to do? Not all services in the runtime should be replaceable using the defaulting mechanism (you can always do this by by replacing the Gradle module on the classpath).

**Reply by lucian-torje-siemens** [reply]
The connector service is created with a catalog service already assigned (I do not need to change it - I just need to change its internal catalog service - maybe it could be injected).

I do need to tweak the catalog service as to filter the output based on a an asset property - in my case the tenant. 
Can you please instruct on how to accomplish such thing in a better way.

see https://github.com/mindsphere/DataSpaceConnector-Fork/pull/37/files#diff-1a13828096f5ea80f6de24a2761a46d006ff7e08fc1e1690f5a9e7cfcf7e079a

![image](https://user-images.githubusercontent.com/102717569/183080665-432f1b02-3e4a-4730-9fc7-99e2d6df58a3.png)

**Reply by jimmarino** [reply]
I don't think this filtering should not be done here. I would handle that in the ContractDefinitionStore where you can generate contract definitions on the fly for the participant agent with the proper selector expression to pass to the AssetIndex.

**Reply by lucian-torje-siemens** [reply]
Did I got it right ?
- the contract offer query is used by contract offer service
- the contract offer creates the participant
![image](https://user-images.githubusercontent.com/102717569/183085791-857d93f9-6a32-4c48-aaf0-9d05fb47129b.png)

-the participant has the claims
![image](https://user-images.githubusercontent.com/102717569/183086986-f2b27958-0ccb-4294-a112-9edb1031d167.png)

- the contract definition service gets called with the participant 
![image](https://user-images.githubusercontent.com/102717569/183087168-199a9760-ddc1-413c-85f2-36e2108dcefa.png)

- contract definition service calls contract definitions store (the persistence part)
![image](https://user-images.githubusercontent.com/102717569/183087452-01f550d2-f641-4ad4-8bb3-5286e68c740e.png)

It looks like only on contract definition service we could filter since the store is participant agnostic:

![image](https://user-images.githubusercontent.com/102717569/183087956-f0fc6b0b-e789-44b4-a4f6-d8978d25b942.png)

**Reply by jimmarino** [reply]
Not quite. The ContractDefinitionStore should be tenant aware. As I mentioned in previous discussions, this can be achieved through the use of a ThreadLocal context that is set using a JAX-RS filter and used by a custom implementation of the ContractDefinitionStore. This will avoid polluting the runtime execution path with tenant logic and also allow data to be segmented properly.

**Reply by lucian-torje-siemens** [reply]
Got it
- create a JAX-RS filter - save the token in a thread local storage
- create custom contractdefinition store and read from the thread local storage
Thank you @jimmarino

**Reply by lucian-torje-siemens** [reply]
@jimmarino The thread local approach has one big risk - if the thread changes (e.g. async work) than the value is lost/used in the wrong thread.

What would be the probability for something like this to happen in the future - as far as I can tell the all flow uses a single thread today, question is if it is going to be the same in the future ?

If the thread is reused (pooled) - we can simple clean it up after usage.

**Reply by jimmarino** [reply]
I don't follow what the concern is. The threadlocal approach would work exactly as in any managed Java container environment (c.f., the Java/Jakarta EE work manager, tracing libraries, etc., for example) - you need to make sure context information is propagated to different threads. For synchronous call blocks, set/unset the threadlocal. If async work is being done, propagate the information as part of an entity such as the `TransferProcess`.

**Reply by lucian-torje-siemens** [reply]
The missing piece of the puzzle was InheritableThreadLocal - see https://onecompiler.com/java/3ych24tw8

For me it still feels more natural to have a list of evaluate access policy filters - a list to which it should be possible to add more filters from custom code; nevertheless the thread-local storage approach will solve my problem. Thank you @jimmarino .

**Reply by jimmarino** [reply]
I would not do that since it has the potential for unintended side effects and will not work in all cases. 

For synchronous code paths, set and clear the ThreadLocal at the ingress point using a filter and finally block. For async requests, attach relevant context to the entity or use your own thread dispatching mechanism like the Java WorkManager API. 

There are examples of these approaches in the EDC codebase - .cf local transactions and metrics. 

The approach I outlined also avoids rippling changes throughout the codebase that involve tunneling context-related data (e.g. 'extra' filters).

**Reply by lucian-torje-siemens** [reply]
Hello @ndr-brt @jimmarino,
I'm trying to replace the default InMemoryAssetIndex with my own custom implementation, unfortunately it is still using the default one marked with isDefault  = true. What should be the right way to do this?

![image](https://user-images.githubusercontent.com/102717569/185318944-c361d913-876c-485a-89d8-761ed482a7cc.png)

**Reply by ndr-brt** [reply]
@lucian-torje-siemens you can find the documentation about dependency resolution here: https://eclipse-dataspaceconnector.github.io/DataSpaceConnector/#/developer/dependency_resolution

if the documentation is not clear enough or there's something that's not working correctly, feel free to open an issue
